### PR TITLE
Datatypes

### PR DESCRIPTION
**Pull request by chasehensel**
Not sure how I ended up with a detached head, but this is the one to review. I'll rebase and clean up everything @awans

### CODE REVIEW

File: internal/api/parse.go [hunk 1]
```diff
@@ -76,15 +76,16 @@ func (p Parser) parseNestedConnect(parentBinding db.Binding, data map[string]int
...
+	m, err := p.tx.GetModelByID(parentBinding.Dual().ModelID())
 	if err != nil {
 		return
 	}
 	// this should be a separate method
 	var uq UniqueQuery
 	for k, v := range data {
 		var val interface{}
-		val, err = m.AttributeByName(k).ParseFromJson(v)
+		d := m.AttributeByName(k).Datatype
```
Comment (awans) [inline internal/api/parse.go:87]: Nice! I love seeing the improved encapsulation.

File: internal/datatypes/core_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package datatypes
+
+import (
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+var datatypeTests = []struct {
```
Comment (awans) [inline internal/datatypes/core_test.go:9]: oh nice! I didn't expect you'd be able to get them all as data. But yes, much clearer.

File: internal/runtime/runtime.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+type Runtime int64
+
+const (
+	Golang Runtime = iota
+	Javascript
+	Starlark
+)
+
+func CallFunc(name string, r Runtime, code string, arg interface{}) (interface{}, error) {
```
Comment (awans) [inline internal/runtime/runtime.go:18]: I'm noticing the awkwardness of "go" vs the "interpreted" languages -- one needing the name (or some kind of handle) and one needing the actual code text. I wonder if we could basically box that difference behind an interface -- maybe something like FunctionHandle.

```
type FunctionHandle interface {
Invoke(arg) arg
}

type GoFunctionHandle struct {
name string
}

type StarlarkFunctionHandle struct {
code string
}
```

Comment (awans) [reply internal/runtime/runtime.go:18]: But I can't tell if that's a bit overdesigned for where we are now.

Comment (awans) [reply internal/runtime/runtime.go:18]: In the medium term, I'd like to see modules be able to provide their own "runtimes". But I'm cool with punting on that until we have more uses of them built out.

Comment (chasehensel) [reply internal/runtime/runtime.go:18]: Are you thinking about having the CallFunc get passed in a FunctionHandle from the db?

Comment (awans) [reply internal/runtime/runtime.go:18]: That was the thought, but TBH I'm feeling like maybe we should just punt on it for now and merge this part in as is!

Comment (chasehensel) [reply internal/runtime/runtime.go]: Actually I got it. Just cleaned this up.

**Pull request merged**